File: bakery-pos-api/controllers/userController.js
const db = require("../config/db");

// ========================
// LOGIN
// ========================
exports.login = async (req, res) => {
  try {
    const { email, password } = req.body;

    const [rows] = await db.execute(
      "SELECT id, name, email, role FROM users WHERE email = ? AND password = ?",
      [email, password]
    );

    if (rows.length === 0) {
      return res.status(401).json({
        success: false,
        message: "Email atau password salah",
      });
    }

    res.json({
      success: true,
      message: "Login berhasil",
      user: rows[0],
    });
  } catch (error) {
    console.error("ERROR LOGIN:", error.message);

    res.status(500).json({
      success: false,
      message: "Gagal login",
      error_detail: error.message,
    });
  }
};

// ========================
// GET ALL USERS
// ========================
exports.getAllUsers = async (req, res) => {
  try {
    const [rows] = await db.execute(
      "SELECT id, name, email, role FROM users ORDER BY id DESC"
    );

    res.json({
      success: true,
      data: rows,
    });
  } catch (error) {
    console.error("ERROR GET USERS:", error.message);

    res.status(500).json({
      success: false,
      message: "Gagal mengambil data user",
      error_detail: error.message,
    });
  }
};

// ========================
// GET USER BY ID
// ========================
exports.getUserById = async (req, res) => {
  try {
    const { id } = req.params;

    const [rows] = await db.execute(
      "SELECT id, name, email, role FROM users WHERE id = ?",
      [id]
    );

    if (rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: "User tidak ditemukan",
      });
    }

    res.json({
      success: true,
      data: rows[0],
    });
  } catch (error) {
    console.error("ERROR GET USER:", error.message);

    res.status(500).json({
      success: false,
      message: "Gagal mengambil user",
      error_detail: error.message,
    });
  }
};

// ========================
// CREATE USER
// ========================
exports.createUser = async (req, res) => {
  try {
    const { name, email, password, role } = req.body;

    // cek email
    const [checkEmail] = await db.execute(
      "SELECT * FROM users WHERE email = ?",
      [email]
    );

    if (checkEmail.length > 0) {
      return res.status(400).json({
        success: false,
        message: "Email sudah digunakan",
      });
    }

    await db.execute(
      "INSERT INTO users (name, email, password, role) VALUES (?, ?, ?, ?)",
      [name, email, password, role]
    );

    res.json({
      success: true,
      message: "User berhasil dibuat",
    });
  } catch (error) {
    console.error("ERROR CREATE USER:", error.message);

    res.status(500).json({
      success: false,
      message: "Gagal membuat user",
      error_detail: error.message,
    });
  }
};

// ========================
// UPDATE USER
// ========================
exports.updateUser = async (req, res) => {
  try {
    const { id } = req.params;

    const { name, email, role, password } = req.body;

    // kalau password diisi
    if (password && password !== "") {
      await db.execute(
        "UPDATE users SET name = ?, email = ?, role = ?, password = ? WHERE id = ?",
        [name, email, role, password, id]
      );
    } else {
      await db.execute(
        "UPDATE users SET name = ?, email = ?, role = ? WHERE id = ?",
        [name, email, role, id]
      );
    }

    res.json({
      success: true,
      message: "User berhasil diperbarui",
    });
  } catch (error) {
    console.error("ERROR UPDATE USER:", error.message);

    res.status(500).json({
      success: false,
      message: "Gagal update user",
      error_detail: error.message,
    });
  }
};

// ========================
// DELETE USER
// ========================
exports.deleteUser = async (req, res) => {
  try {
    const { id } = req.params;

    await db.execute(
      "DELETE FROM users WHERE id = ?",
      [id]
    );

    res.json({
      success: true,
      message: "User berhasil dihapus",
    });
  } catch (error) {
    console.error("ERROR DELETE USER:", error.message);

    res.status(500).json({
      success: false,
      message: "Gagal menghapus user",
      error_detail: error.message,
    });
  }
};

File: lib/controller/admin/user_controller.dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
// Sesuaikan path
import 'package:tugas_akhir/models/user.dart';
import 'package:tugas_akhir/api service/api_service.dart';

class UserController extends GetxController {
  final RxList<User> users = <User>[].obs;
  final RxList<User> filteredUsers = <User>[].obs;

  final RxBool isLoading = true.obs;
  final RxBool isSubmitting = false.obs;

  final TextEditingController searchController = TextEditingController();

  // Pagination Table (Desktop)
  final RxInt currentPage = 1.obs;
  final int itemsPerPage = 10;

  @override
  void onInit() {
    super.onInit();
    searchController.addListener(_onSearchChanged);
    fetchUsers();
  }

  @override
  void onClose() {
    searchController.removeListener(_onSearchChanged);
    searchController.dispose();
    super.onClose();
  }

  void _onSearchChanged() {
    searchUsers(searchController.text);
  }

  // ========================
  // GET DATA
  // ========================
  Future<void> fetchUsers() async {
    try {
      isLoading.value = true;
      final data = await ApiService.getAllUsers();
      users.assignAll(data);
      filteredUsers.assignAll(data);
      currentPage.value = 1;
    } catch (e) {
      Get.snackbar(
        'Error',
        'Gagal mengambil data user: $e',
        backgroundColor: Colors.red,
        colorText: Colors.white,
      );
    } finally {
      isLoading.value = false;
    }
  }

  // ========================
  // SEARCH & FILTER
  // ========================
  void searchUsers(String query) {
    if (query.isEmpty) {
      filteredUsers.assignAll(users);
    } else {
      final lowerQuery = query.toLowerCase();
      filteredUsers.assignAll(
        users.where((user) {
          return user.name.toLowerCase().contains(lowerQuery) ||
              user.email.toLowerCase().contains(lowerQuery) ||
              user.role.toLowerCase().contains(lowerQuery);
        }).toList(),
      );
    }
    currentPage.value = 1;
  }

  // ========================
  // PAGINATION LOGIC
  // ========================
  int get totalPages => (filteredUsers.isEmpty)
      ? 1
      : (filteredUsers.length / itemsPerPage).ceil();

  List<User> get paginatedUsers {
    final startIndex = (currentPage.value - 1) * itemsPerPage;
    return filteredUsers.skip(startIndex).take(itemsPerPage).toList();
  }

  void nextPage() {
    if (currentPage.value < totalPages) currentPage.value++;
  }

  void previousPage() {
    if (currentPage.value > 1) currentPage.value--;
  }

  void setPage(int page) {
    if (page >= 1 && page <= totalPages) currentPage.value = page;
  }

  // ========================
  // CRUD OPERATIONS
  // ========================
  Future<void> createUser({
    required String name,
    required String email,
    required String role,
    required String password,
  }) async {
    try {
      isSubmitting.value = true;
      final newUser = User(
        name: name,
        email: email,
        role: role,
        password: password,
      );

      bool success = await ApiService.createUser(newUser);
      if (success) {
        Get.back(); // Tutup dialog
        Get.snackbar(
          'Sukses',
          'Akun berhasil ditambahkan',
          backgroundColor: Colors.green,
          colorText: Colors.white,
        );
        fetchUsers();
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        e.toString(),
        backgroundColor: Colors.red,
        colorText: Colors.white,
      );
    } finally {
      isSubmitting.value = false;
    }
  }

  // Hanya mengubah nama, email, dan role.
  // Password lama TIDAK PERNAH diminta atau dikirim di sini.
  Future<void> updateUser({
    required int id,
    required String name,
    required String email,
    required String role,
  }) async {
    try {
      isSubmitting.value = true;
      final updatedUser = User(id: id, name: name, email: email, role: role);

      bool success = await ApiService.updateUser(updatedUser);
      if (success) {
        Get.back(); // Menutup dialog setelah sukses
        Get.snackbar(
          'Sukses',
          'Akun berhasil diperbarui',
          backgroundColor: Colors.green,
          colorText: Colors.white,
        );
        fetchUsers(); // Refresh data table/list
      } else {
        // Jika server merespon status 200 tapi operasi bernilai gagal secara logika bisnis
        Get.snackbar(
          'Gagal',
          'Gagal memperbarui data. Periksa kembali hak akses Anda.',
          backgroundColor: Colors.amber.shade800,
          colorText: Colors.white,
        );
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        e.toString(),
        backgroundColor: Colors.red,
        colorText: Colors.white,
      );
    } finally {
      isSubmitting.value = false;
    }
  }

  // ========================
  // RESET PASSWORD
  // ========================
  // Alur khusus Admin: Admin menimpa password lama dengan password baru.
  // Frontend tidak pernah membaca / menampilkan password lama,
  // dan tidak melakukan hashing (hashing dilakukan backend).
  Future<void> resetPassword({
    required int id,
    required String newPassword,
  }) async {
    try {
      isSubmitting.value = true;
      bool success = await ApiService.resetPassword(
        id: id,
        password: newPassword,
      );
      if (success) {
        Get.back();
        Get.snackbar(
          'Sukses',
          'Password akun berhasil direset',
          backgroundColor: Colors.green,
          colorText: Colors.white,
        );
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        e.toString(),
        backgroundColor: Colors.red,
        colorText: Colors.white,
      );
    } finally {
      isSubmitting.value = false;
    }
  }

  Future<void> deleteUser(int id) async {
    try {
      isSubmitting.value = true;
      bool success = await ApiService.deleteUser(id);
      if (success) {
        Get.back();
        Get.snackbar(
          'Sukses',
          'Akun berhasil dihapus',
          backgroundColor: Colors.green,
          colorText: Colors.white,
        );
        fetchUsers();
      }
    } catch (e) {
      Get.snackbar(
        'Error',
        'Gagal menghapus akun: $e',
        backgroundColor: Colors.red,
        colorText: Colors.white,
      );
    } finally {
      isSubmitting.value = false;
    }
  }
}
